dali/operators/geometry/transform_base_op.h

```diff
@@ -0,0 +1,185 @@
+
+namespace dali {
+
+template <int... values>
+using dims = std::integer_sequence<int, values...>;
+
+template <typename T, int ndim>
+using affine_mat_t = mat<ndim+1, ndim+1, T>;
+
+/**
+ * @brief Base CRTP class for affine transform generators.
+ * The matrix definition comes from the actual TransformImpl implementation.
+ * As with any CRTP-based system, any non-private method can be shadowed by the TransformImpl class.
+ */
+template <typename Backend, typename TransformImpl>
+class TransformBaseOp : public Operator<Backend> {
+ public:
+  explicit TransformBaseOp(const OpSpec &spec)
+      : Operator<Backend>(spec),
+        spec_(spec),
+        nsamples_(spec.GetArgument<int>("batch_size")) {
+  }
+
+  bool CanInferOutputs() const override { return true; }
+
+  TransformImpl &This() noexcept { return static_cast<TransformImpl&>(*this); }
+  const TransformImpl &This() const noexcept { return static_cast<const TransformImpl&>(*this); }
+
+ protected:
+  void CheckInputDimensionality(const workspace_t<CPUBackend> &ws) {
+    if (has_input_) {
+      auto in_t_ndim = input_transform_ndim(ws);
+      DALI_ENFORCE(in_t_ndim == ndim_,
+        make_string(
+          "Unexpected number of dimensions: expected ", ndim_, 
+          " dimensions but the input transform has ", in_t_ndim));
+    }
+  }
+
+  bool SetupImpl(std::vector<OutputDesc> &output_descs, const workspace_t<Backend> &ws) override {
+    has_input_ = ws.template NumInput() > 0;
+    if (has_input_) {
+      auto &input = ws.template InputRef<Backend>(0);
+      const auto &shape = input.shape();
+      DALI_ENFORCE(is_uniform(shape), "All matrices must have the same shape.");
+      DALI_ENFORCE(input.type().id() == dtype_,
+        make_string("Unexpected input data type. Expected ", dtype_, ", got: ", input.type().id()));
+
+      DALI_ENFORCE(shape.sample_dim() == 2 &&
+                   shape.size() > 0 &&
+                   shape[0][1] == (shape[0][0] + 1),
+        make_string(
+          "The input, if provided, is expected to be a 2D tensor with dimensions "
+          "(ndim, ndim+1) representing an affine transform. Got: ", shape));
+      nsamples_ = shape.num_samples();
+    }
+
+    This().ProcessArgs(spec_, ws);
+    CheckInputDimensionality(ws);
+
+    output_descs.resize(1);  // only one output
+    output_descs[0].type = TypeTable::GetTypeInfo(dtype_);
+    output_descs[0].shape = uniform_list_shape(nsamples_, {ndim_, ndim_+1});
+    return true;
+  }
+
+  template <typename T>
+  void RunImplTyped(workspace_t<Backend> &ws, dims<>) {
+    DALI_FAIL(make_string("Unsupported number of dimensions ", ndim_));
+  }
+
+  template <typename T, int ndim, int... ndims>
+  void RunImplTyped(workspace_t<Backend> &ws, dims<ndim, ndims...>) {
+    if (ndim_ != ndim) {
+      RunImplTyped<T>(ws, dims<ndims...>());
+      return;
+    }
+
+    auto &out = ws.template OutputRef<Backend>(0);
+    out.SetLayout("");  // TODO(janton): Decide what layout we want for transforms
```

Comment:
I think we can live with no layout for plain matrices - unless we want to somehow stress that they're row-major - then maybe "rc" (row col).